Add tests

emilcheva opened the pull request:
Mocked graphql queries with MSW and added tests accordingly

emilcheva committed:
Add tests

Mocked graphql queries with MSW and added tests accordingly

steliyan commented on package.json:
```diff
@@ -51,14 +51,16 @@
   "devDependencies": {
     "@testing-library/jest-dom": "^4.2.4",
-    "@testing-library/react": "^9.3.2",
-    "@testing-library/user-event": "^7.1.2",
+    "@testing-library/react": "^12.1.2",
+    "@testing-library/user-event": "^13.5.0",
     "autoprefixer": "^10.4.1",
     "eslint-config-prettier": "^8.3.0",
     "eslint-plugin-prettier": "^4.0.0",
     "lodash.debounce": "^4.0.8",
+    "msw": "^0.36.7",
     "postcss": "^8.4.5",
-    "prettier": "^2.5.1"
+    "prettier": "^2.5.1",
+    "react-test-renderer": "^17.0.2"
```
NIT: I would lock the dependency versions. If you decide to do it, do it in a separate PR.

emilcheva replied:
Ok, will resolve with another PR

steliyan commented on src/ApolloClient.js:
```diff
@@ -0,0 +1,15 @@
+import { ApolloClient, HttpLink, InMemoryCache } from '@apollo/client';
+
+const cache = new InMemoryCache();
+
+const link = new HttpLink({
+  uri: 'https://cat-wiki-apollo-server.herokuapp.com',
+  cache: new InMemoryCache()
+});
```
Why 2 separate cache objects? It's bound to end in some weird UI state error.

emilcheva replied:
Fixed

steliyan commented on src/components/__tests__/autocomplete.test.js:
```diff
@@ -0,0 +1,88 @@
+
+const autoCompleteSetup = () => {
+  render(
+    <ApolloProvider client={client}>
```
The way the client is passed to the `ApolloProvider` the `client` is shared between the tests in this file (test suite). Is this expected?

emilcheva replied:
yep, intended :)

steliyan replied:
Why?

steliyan commented on src/mocks/setupServer.js:
```diff
@@ -0,0 +1,17 @@
+import { server } from './server';
+
+export const setupServer = function () {
```
Why not an arrow function?

steliyan commented on src/mocks/handlers.js:
```diff
@@ -0,0 +1,90 @@
+export const handlers = [
+  // Handles a "GetBreeds" query
+  graphql.query('getBreeds', (req, res, ctx) => {
+    // const { limit } = req.variables;
```
Leftover?

steliyan commented on src/components/autocomplete.js:
```diff
@@ -8,7 +8,7 @@ import styled from '@emotion/styled';
 
 export const GET_BREED = gql`
-  query getBreed($breedName: String!) {
+  query getAutoCompleteBreed($breedName: String!) {
```
Why not `getBreedByName`?

steliyan commented on src/components/autocomplete.js:
```diff
@@ -26,18 +26,18 @@ const Autocomplete = () => {
+    debounce((text) => {
       getCatBreeds({
-        variables: { breedName: query }
+        variables: { breedName: text }
```
NIT: You can rename `text` to `breedName` and simplify the obj construction.

steliyan commented on src/components/autocomplete.js:
```diff
@@ -26,18 +26,18 @@ const Autocomplete = () => {
 
-    debouncedChangeFilter();
-  }
+  useEffect(() => {
+    debouncedApiCall(input);
+  }, [input, debouncedApiCall]);
```
It's a good idea to clean-up in case this component gets unmounted.

You can use `.cancel`: https://lodash.com/docs/4.17.15#debounce

steliyan commented on src/components/autocomplete.js:
```diff
@@ -26,18 +26,18 @@ const Autocomplete = () => {
+  }, [input, debouncedApiCall]);
 
   useEffect(() => {
     if (data && !loading) {
```
I think this may be true if the current query returns an error, then `data` would hold the previous result. Not sure if this is what we want.

steliyan commented on src/components/autocomplete.js:
```diff
@@ -46,16 +46,10 @@ const Autocomplete = () => {
-
   const keyDownHandler = (e) => {
     // enter key
     if (e.keyCode === 13) {
```
`const ENTER_KEY_CODE = 13` and the comment can safely be deleted. :)

steliyan commented on src/components/breed-detail.js:
```diff
@@ -31,7 +31,16 @@ const BreedDetail = ({ breed }) => {
         <Row>
           <Col xs="auto" className="mb-4 me-lg-5">
-            <ImageWrapper width="270" height="270" src={breedImage[0].url} alt={name} />
+            {breedImage.length ? (
+              <ImageWrapper width="270" height="270" src={breedImage[0]?.url} alt={name} />
+            ) : (
+              <ImageWrapper
+                width="270"
+                height="270"
+                src="https://images.unsplash.com/photo-1577349516274-37ff88a53627?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1035&q=80"
+                alt={name}
+              />
+            )}
```
Both `ImageWrapper`s are basically the same, why not just specify the appropriate prop?

steliyan commented on src/components/breed-detail.js:
```diff
@@ -77,14 +86,26 @@ const BreedDetail = ({ breed }) => {
+            </Row>
+          </>
+        ) : (
+          ''
```
Why not `nuill`?

steliyan commented on src/components/breed-detail.js:
```diff
@@ -77,14 +86,26 @@ const BreedDetail = ({ breed }) => {
+            <h2>Other photos</h2>
+            <Row>
+              {breedImage.map((image, index) => (
+                <Col sm="6" lg="3" className="mb-4" key={index}>
```
Don't use `index` for a `key`: https://reactjs.org/docs/lists-and-keys.html

steliyan commented on src/mocks/setupServer.js:
```diff
@@ -0,0 +1,17 @@
+export const setupServer = function () {
+  beforeAll(() => {
+    server.listen({
+      onUnhandledRequest: 'warn'
```
Why is that needed? I would advise treating every warning as an error in the tests as this may uncover potential problems with the production code (or the mocks).

steliyan commented on src/components/__tests__/autocomplete.test.js:
```diff
@@ -0,0 +1,88 @@
+};
+
+describe('Testing Autocomplete w getBreedsByName query', () => {
+  setupServer();
```
This gets executed once per `describe`. Most likely, you want to do it before each test. If not, `beforeAll` can be used scoped in the `describe`. Either way, it's more explicit and easier for the reader.

steliyan commented on src/components/__tests__/autocomplete.test.js:
```diff
@@ -0,0 +1,88 @@
+    userEvent.type(screen.getByPlaceholderText(/enter your breed/i), 'po');
+
+    await waitFor(() => {
+      expect(screen.queryAllByTestId('suggestions-list')[0]).toBeInTheDocument();
```
Seems redundant given the assertion below.

steliyan commented on src/components/__tests__/autocomplete.test.js:
```diff
@@ -0,0 +1,88 @@
+    screen.getByPlaceholderText(/enter your breed/i).focus();
+    fireEvent.keyDown(document.activeElement || document.body, {
+      key: 'Enter',
+      keyCode: 13,
+      which: 13
```
I think `shich` is redundant here.

steliyan commented on src/components/__tests__/autocomplete.test.js:
```diff
@@ -0,0 +1,88 @@
+      expect(screen.queryAllByTestId('suggestions-list')[0]).toBeInTheDocument();
+    });
+    screen.getByPlaceholderText(/enter your breed/i).focus();
+    fireEvent.keyDown(document.activeElement || document.body, {
```
Why the `||` here?

steliyan commented on src/components/__tests__/components.test.js:
```diff
@@ -0,0 +1,37 @@
+      </ApolloProvider>
+    );
+
+    const breedName = await screen.findByText('Munchkin');
+    expect(breedName).toBeTruthy();
+    const temperament = await screen.findByText(/agile, easy going, intelligent, playful/i);
+    expect(temperament).toBeTruthy();
```
Can we use `toBeInTheDocument` instead of `toBeTruthy`?

steliyan commented on src/components/__tests__/components.test.js:
```diff
@@ -0,0 +1,37 @@
+import { setupServer } from '../../mocks/setupServer';
+import Breed from '../../pages/breed';
+
+describe('Testing TopBreeds w getBreeds query', () => {
```
`Testing` in the title, why? :)

steliyan commented on src/components/__tests__/query-result.test.js:
```diff
@@ -0,0 +1,23 @@
+describe('Query Result', () => {
+  afterEach(cleanup);
+
+  it('renders loading state', async () => {
+    render(<QueryResult loading={true} />);
+    expect(screen.getByTestId(/spinner/i)).toBeInTheDocument();
+  });
+
+  it('renders No Data message', async () => {
+    render(<QueryResult loading={false} />);
+    expect(screen.getByText(/no data to show/i)).toBeInTheDocument();
+  });
+
+  it('renders Error', async () => {
+    render(<QueryResult error={new ApolloError('something erroneous happened')} />);
+    expect(screen.getByText((content) => content.startsWith('ERROR'))).toBeInTheDocument();
```
No need for `async`.

steliyan commented on src/components/__tests__/query-result.test.js:
```diff
@@ -0,0 +1,23 @@
+import React from 'react';
+import { render, cleanup, screen } from '@testing-library/react';
+import QueryResult from '../query-result';
```
That's a weird component, given how you use it. Why the null-check in the children?

emilcheva committed:
Resolving PR comments

emilcheva committed:
Fix for navigation with dropdown arrows on the suggestions list and the 'not found' text being shown on empty input

Implemented "focus trap" for the dropdown items so that there's always one selected , fixed the 'not found' msg appearance and updated tests accordingly

steliyan commented on src/ApolloClient.js:
```diff
@@ -4,7 +4,7 @@ const cache = new InMemoryCache();
 const link = new HttpLink({
   uri: 'https://cat-wiki-apollo-server.herokuapp.com',
-  cache: new InMemoryCache()
+  cache: cache
```
`cache: cache` -> `cache`.

steliyan commented on src/components/autocomplete.js:
```diff
@@ -63,7 +68,7 @@ const Autocomplete = () => {
     }
     // down arrow
     else if (e.keyCode === 40) {
```
Same comment as for the `ENTER` key.

steliyan commented on src/mocks/handlers.js:
```diff
@@ -73,17 +73,26 @@ export const handlers = [
+        })
+      );
+    }
   })
```
Should there be `else if` or just `else`?

steliyan commented on src/components/breed-detail.js:
```diff
@@ -31,16 +31,18 @@ const BreedDetail = ({ breed }) => {
+              width="270"
+              height="270"
+              src={
+                breedImage.length
+                  ? breedImage[0]?.url
+                  : 'https://images.unsplash.com/photo-1577349516274-37ff88a53627?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1035&q=80'
```
NIT: You can extract this in a separate func.

steliyan commented on src/components/breed-detail.js:
```diff
@@ -90,8 +92,8 @@ const BreedDetail = ({ breed }) => {
-              {breedImage.map((image, index) => (
-                <Col sm="6" lg="3" className="mb-4" key={index}>
+              {breedImage.map((image) => (
+                <Col sm="6" lg="3" className="mb-4" key={breedImage.id}>
```
Should this be `breedImage.id` or `image.id`?

steliyan commented on src/components/breed-detail.js:
```diff
@@ -90,8 +92,8 @@ const BreedDetail = ({ breed }) => {
           <>
```
Does the `.length` condition work if `0` items exist?

steliyan commented on src/components/autocomplete.js:
```diff
@@ -26,36 +26,35 @@ const Autocomplete = () => {
+    debouncedApiCall(input);
+
+    return () => {
+      debouncedApiCall.cancel();
```
Does this work? Shouldn't the return result of the callback get cancelled?

steliyan commented on src/components/autocomplete.js:
```diff
@@ -82,17 +81,22 @@ const Autocomplete = () => {
         <InputSearch id="search">
-          <Link to={`/breed/${input}`}>
-            <Search style={{ fill: colors.dark }} />
+          <Link to={input ? `/breed/${input}` : ''}>
```
Do we need a `Link` component in the `input` is falsy?

steliyan commented on src/components/autocomplete.js:
```diff
@@ -82,17 +81,22 @@ const Autocomplete = () => {
-            <Search style={{ fill: colors.dark }} />
+          <Link to={input ? `/breed/${input}` : ''}>
+            <Search
+              style={{
+                fill: input ? colors.dark : colors.darkGrey,
+                cursor: input ? 'pointer' : 'not-allowed'
+              }}
```
The style object is new on each rerender. Generally, we don't use inline styles.

steliyan commented on src/mocks/handlers.js:
```diff
@@ -0,0 +1,98 @@
+import { graphql } from 'msw';
+
+export const handlers = [
+  // Handles a "GetBreeds" query
```
Pretty redundant. Most likely will get outdated if something gets fixed down the file.

steliyan commented on src/mocks/handlers.js:
```diff
@@ -0,0 +1,98 @@
+          ]
+        })
+      );
+    } else if (breedName === 'brv') {
+      return res(
+        ctx.data({
+          getBreedsByName: []
+        })
+      );
+    }
+  })
```
Same question as before, should this be just `else`?

emilcheva committed:
PR fixes and new test to clarify input is debounced

We ensure that request query variables contain the whole typing, not single letters

emilcheva committed:
Ensure debouncing test works

Need to ensure that variables sent does not contain single letters but whole typed word

steliyan commented on src/components/__tests__/autocomplete.test.js:
```diff
@@ -6,6 +6,7 @@ import { client } from '../../ApolloClient';
+import { jssPreset } from '@material-ui/core';
 
 const ENTER_KEY_CODE = 13;
 const REQUESTS_MAP = new Map();
```
That's not really a "const", so better to be `requestsMap`.

steliyan commented on src/components/__tests__/autocomplete.test.js:
```diff
@@ -111,10 +112,13 @@ describe('Autocomplete w getBreedsByName query', () => {
-    });
+    for (let [, value] of REQUESTS_MAP.entries()) {
+      getAllRequestVariables.push(value['breedName']);
+    }
```
This can be rewritten as:
```const getAllRequestVariables = REQUESTS_MAP.values().map(({ breedName }) => breedName)```.

steliyan commented on src/components/__tests__/autocomplete.test.js:
```diff
@@ -111,10 +112,13 @@ describe('Autocomplete w getBreedsByName query', () => {
+      getAllRequestVariables.push(value['breedName']);
+    }
+
+    expect(getAllRequestVariables.some((variable) => variable === 'p')).toBe(false);
+    expect(getAllRequestVariables.some((variable) => variable === 'pers')).toBe(true);
```
`toBe(false)` -> `toBeFalsy`
`toBe(true)` -> `toBeTruthy`

steliyan replied:
If the API call is made twice, the test would pass again.

steliyan commented on src/components/__tests__/autocomplete.test.js:
```diff
@@ -111,10 +112,13 @@ describe('Autocomplete w getBreedsByName query', () => {
+      getAllRequestVariables.push(value['breedName']);
+    }
+
+    expect(getAllRequestVariables.some((variable) => variable === 'p')).toBe(false);
```
Why don't we check for `pe` and `per` then?

emilcheva committed:
PR comments resolutions

Added more assertions for variables that should not be sent if debouncing works

steliyan commented on src/components/__tests__/autocomplete.test.js:
```diff
@@ -112,13 +111,11 @@ describe('Autocomplete w getBreedsByName query', () => {
-
-    expect(getAllRequestVariables.some((variable) => variable === 'p')).toBe(false);
-    expect(getAllRequestVariables.some((variable) => variable === 'pers')).toBe(true);
+    expect(getAllRequestVariables.some((variable) => variable === 'p')).toBeFalsy;
+    expect(getAllRequestVariables.some((variable) => variable === 'pe')).toBeFalsy;
+    expect(getAllRequestVariables.some((variable) => variable === 'per')).toBeFalsy;
+    expect(getAllRequestVariables.some((variable) => variable === 'pers')).toBeTruthy;
```
`toBeFalsy`/`toBeTruthy` are functions, but they are not invoked.

steliyan replied:
Can we just do `expect(getAllRequestVariables).toEqual(['pers'])`?

steliyan commented on src/components/__tests__/autocomplete.test.js:
```diff
@@ -112,13 +111,11 @@ describe('Autocomplete w getBreedsByName query', () => {
     });
 
-    let getAllRequestVariables = [];
+    const getAllRequestVariables = [requestsMap.values()].map(({ breedName }) => breedName);
```
Shouldn't the `values()` get spreaded?

emilcheva committed:
PR comments resolves

ApolloClient is caching, we need to reinit it for each individual test

steliyan commented on src/ApolloClient.js:
```diff
@@ -1,15 +1,13 @@
+export const CREATE_CLIENT = new ApolloClient({
+  link,
+  cache
 });
```
This should be a function, not a client instance.

steliyan commented on src/ApolloClient.js:
```diff
@@ -1,15 +1,13 @@
 const link = new HttpLink({
   uri: 'https://cat-wiki-apollo-server.herokuapp.com',
-  cache: cache
+  cache
```
NIT: You can just inline the cache creation here.

steliyan commented on src/ApolloClient.js:
```diff
@@ -1,15 +1,13 @@
-  link
+export const CREATE_CLIENT = new ApolloClient({
+  link,
+  cache
```
Is this needed here?

steliyan commented on src/components/__tests__/autocomplete.test.js:
```diff
@@ -2,28 +2,40 @@ import React from 'react';
+});
+
+afterAll(() => {
+  server.events.removeAllListeners();
```
Better to remove only the registered listener, not all.

steliyan commented on src/components/__tests__/autocomplete.test.js:
```diff
@@ -2,28 +2,40 @@ import React from 'react';
+
+afterEach(() => {
+  requestsMap.clear();
+  CREATE_CLIENT.resetStore();
```
This would be redundant once the client is instanced per test, not per test suite.

emilcheva committed:
ApolloClient reinit on test case

Creation of new Apollo client instance done in a func and updated the code accordingly

emilcheva committed:
Add another test that verifies fast typing does not sends requests

Typing faster than debounce time should not result in sending requests

steliyan commented on src/components/__tests__/autocomplete.test.js:
```diff
@@ -128,4 +128,27 @@ describe('Autocomplete w getBreedsByName query', () => {
   });
+
+  it('should not send request for fast typing ( keystroke faster than 500ms )', async () => {
+    jest.useFakeTimers('modern');
```
This should be default for jest 27.

steliyan commented on src/components/__tests__/autocomplete.test.js:
```diff
@@ -128,4 +128,27 @@ describe('Autocomplete w getBreedsByName query', () => {
+    act(() => {
+      jest.advanceTimersByTime(499);
+    });
+    expect([...requestsMap.values()].length).toEqual(0);
```
You can use `toHaveLength`:  https://jestjs.io/docs/expect#tohavelengthnumber

steliyan commented on src/components/__tests__/autocomplete.test.js:
```diff
@@ -128,4 +128,27 @@ describe('Autocomplete w getBreedsByName query', () => {
+    autoCompleteSetup();
+
+    const searchInput = screen.getByPlaceholderText(/enter your breed/i);
+    let typed = [];
```
Do we need this?

steliyan commented on src/components/__tests__/autocomplete.test.js:
```diff
@@ -128,4 +128,27 @@ describe('Autocomplete w getBreedsByName query', () => {
+      typed.push(searchInput.value);
+    });
+    const getAllRequestVariables = [...requestsMap.values()].map(({ breedName }) => breedName);
+    expect(getAllRequestVariables).toEqual([...typed]);
```
Why the spread here?

steliyan commented on src/components/__tests__/autocomplete.test.js:
```diff
@@ -128,4 +128,27 @@ describe('Autocomplete w getBreedsByName query', () => {
+    });
+    expect([...requestsMap.values()].length).toEqual(0);
+    act(() => {
+      jest.advanceTimersByTime(501);
```
Should this be `501` or just `1`?

emilcheva committed:
Adjust timer advancement and remove 'typed' array

PR comments resolvements
